### PR TITLE
Auto + Alignment

### PR DESCRIPTION
**Pull request by linglejack06**
- add basic auto paths for drive
- begin implementation of alignment commands

### CODE REVIEW

**Comment by linglejack06** [comment]
@jkleiber do you know why my alignment command isn't working? the auto just tried to point it directly to the left. Is there an issue with the target pose I'm setting?

**Comment by jkleiber** [comment]
@linglejack06 if you could post some advantage scope plots of the key variables that would be helpful in figuring out the root cause

**Comment by linglejack06** [comment]
@jkleiber i think ive got the issue fixed, when i merged main into the code something went wrong. So i reset to before merging and the robot aligns how it should. Path is oddly flipped though starting at blue speaker even though my pathplanner shows it at red speaker, any clue why?

**Comment by jkleiber** [comment]
@linglejack06 this probably has to do with your path planner configuration in the drivetrain subsystem. Also check your sim settings that you're on the alliance you think you're on

**Comment by linglejack06** [comment]
@jkleiber @aidnem could one of you take a look at the auto-alignment (housed in the drivetrain getAutoRotaion method) and the teleop auto-alignment (housed in AlignToTarget command)? The approach I used for teleop helps isolate some logic away from the drivetrain. All that is left for auto now is adding named commands for shooting and intaking, I think.

**Comment by linglejack06** [comment]
@jkleiber i have added a isDriveAligned Method to interface with scoring for auto. But i have a quick question in regard to teleop. I took the route of seperating alignment logic from drive, so its housed in a command that sends goal speeds to drive. With this route, calling is drive aligned adds unnecessary steps (it finds the target heading again and compares it again). These are unecessary because the command will automatically end when the drive is aligned due to the logic inside. 
Could we look into setting up a sequential command like this: align to target -> shoot at target? By using a sequential command, we could avoid calling the isDriveAligned method in teleop, as the scoring subsystem would only run once the alignment command finishes (it's aligned).

**Comment by jkleiber** [comment]
@linglejack06 could we keep alignment logic in drive rather than putting it in commands? 

The reason I don't like to put control logic like this in commands is because then you run into this exact question about auto vs teleop. Ideally the subsystem will manage its own state on if it's aligned or not and can report its status to whatever commands need to know the status

**Comment by linglejack06** [comment]
Yeah I can refactor that

**Comment by aidnem** [comment]
@linglejack06 @jkleiber could we consider using alignState to enable/disable alignment instead of just using alignTarget.NONE? This would allow masher to select an alignTarget without forcing robot to align, until the driver actually pulls the trigger (or presses the button) to start aligning. I feel that this would be an easier solution for the driver, and this is the way it was done on main bot 2024.

**Comment by jkleiber** [comment]
@aidnem that sounds reasonable to me

**Comment by linglejack06** [comment]
@aidnem yeah I can easily add that. I would say we have the driver pull a trigger to align and continue holding, maybe on release of trigger the alignment stops? What do you think of that idea? @jkleiber

**Comment by aidnem** [comment]
If I recall correctly that's how it worked last year as well, I can ask Oliver tonight to make sure

**Comment by linglejack06** [comment]
look into field centric facing angle swerve request for alignment

**Comment by linglejack06** [comment]
@jkleiber ive switched to using the FieldCentricFacingAngle swerve request. no difference that i can tell, but check it out and see which one you like better? @aidnem too

File: src/main/java/frc/robot/constants/FeatureFlags.java
Comment (jkleiber) [inline src/main/java/frc/robot/constants/FeatureFlags.java:5]: We might as well leave this true

File: src/main/java/frc/robot/RobotContainer.java
Comment (jkleiber) [inline src/main/java/frc/robot/RobotContainer.java:318]: Up/Down/Left/Right aren't intuitive alignment targets like speaker/amp/source/endgame. Could we go back to using the intuitive game targets rather than the POV directions?

File: src/main/java/frc/robot/subsystems/drive/PhoenixDrive.java
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/drive/PhoenixDrive.java:277]: This probably should be set via a command rather than always forcing alignment

**Review by jkleiber — COMMENTED** [review]
First pass, I think this is close to being done

File: src/main/java/frc/robot/constants/FeatureFlags.java
Comment (linglejack06) [reply src/main/java/frc/robot/constants/FeatureFlags.java:5]: yeah, i had turned it off as it messes with positioning in sim, so it was throwing off the alignment. but ill put it back on

File: src/main/java/frc/robot/RobotContainer.java
Comment (linglejack06) [reply src/main/java/frc/robot/RobotContainer.java:318]: switched up(speaker) right(amp), left(source), down(endgame), and added right joystick trigger as aligning true

File: src/main/java/frc/robot/subsystems/drive/PhoenixDrive.java
Comment (linglejack06) [reply src/main/java/frc/robot/subsystems/drive/PhoenixDrive.java:277]: that was for testing, updated now

**Review by jkleiber — APPROVED** [review]
approved

**Comment by linglejack06** [comment]
@jkleiber merged main in, needs reapproval

**Review by jkleiber — DISMISSED** [review]

**Review by aidnem — CHANGES_REQUESTED** [review]
Looks fine to me except for the idle thing. If you wouldn't mind explaining that to me (If I'm misunderstanding the code) I'll approve it.

File: src/main/java/frc/robot/subsystems/drive/PhoenixDrive.java
Comment (linglejack06) [reply src/main/java/frc/robot/subsystems/drive/PhoenixDrive.java:208]: @aidenm you're right. That was an issue from merging from main, I'll fix that when I'm back at my computer.

**Comment by linglejack06** [comment]
@jkleiber @aidenm issue fixed just need approval to merge

**Review by aidnem — DISMISSED** [review]
Looks good to me 👍

**Review by jkleiber — APPROVED** [review]

**Comment by aidnem** [comment]
@linglejack06 did you ever add a check that will disable idling when we are trying to align aren't aligned yet?

**Comment by linglejack06** [comment]
I'll add that check soon

File: src/main/java/frc/robot/subsystems/drive/PhoenixDrive.java
Comment (aidnem) [inline src/main/java/frc/robot/subsystems/drive/PhoenixDrive.java:207]: Could we try
```java
if (idling && !(aligning && !isDriveAligned))
```
or something, so that we don't go into X brake when aligning but done moving?
Or will a FieldCentricFacingAngle request not go into X brake?

Comment (linglejack06) [reply src/main/java/frc/robot/subsystems/drive/PhoenixDrive.java:207]: @aidnem i can't find anything in docs, I say we just try out what's here and if it goes into x brake tonight I can update it real quick and then get merged after testing

Comment (aidnem) [reply src/main/java/frc/robot/subsystems/drive/PhoenixDrive.java:207]: @linglejack06 did we ever end up testing this

Comment (linglejack06) [reply src/main/java/frc/robot/subsystems/drive/PhoenixDrive.java:207]: @aidnem no we didnt, i sepnt rest of time trying to fix shooter as one of the motors wasnt working (an electrical issue i think). Justin stayed past nine and I believe shooting works well now. So we can now intake and shoot.

Comment (aidnem) [reply src/main/java/frc/robot/subsystems/drive/PhoenixDrive.java:207]: @linglejack06 Good to know, thanks!

**Comment by jkleiber** [comment]
We should merge this and follow your suggestion @linglejack06

**Comment by linglejack06** [comment]
it is, only thing changed in code tested that isnt here is pid gains and i
had to reverse direction of speaker rotation by subtracting pi from it
(that is in the 14-calibrate-cameras branch)

**Comment by jkleiber** [comment]
@linglejack06 I skimmed over this branch vs the calibrate cameras branch. You're good to merge this since that other branch is serving as the integration branch. Hopefully merging this to main will make the final review of the integration branch easier

**Pull request merged**
